streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala

```diff
@@ -144,27 +141,31 @@ private[streaming] class ReceiverSchedulingPolicy {
     val scheduledExecutors = mutable.Set[String]()
     scheduledExecutors ++= preferredLocation
 
-    val executorWeights = receiverTrackingInfoMap.values.flatMap { receiverTrackingInfo =>
-      receiverTrackingInfo.state match {
-        case ReceiverState.INACTIVE => Nil
-        case ReceiverState.SCHEDULED =>
-          val scheduledExecutors = receiverTrackingInfo.scheduledExecutors.get
-          // The probability that a scheduled receiver will run in an executor is
-          // 1.0 / scheduledLocations.size
-          scheduledExecutors.map(location => location -> (1.0 / scheduledExecutors.size))
-        case ReceiverState.ACTIVE => Seq(receiverTrackingInfo.runningExecutor.get -> 1.0)
-      }
+    val executorWeights = receiverTrackingInfoMap.filter(_._1 != receiverId).values.flatMap {
+      receiverTrackingInfo =>
+        receiverTrackingInfo.state match {
+          case ReceiverState.INACTIVE => Nil
+          case ReceiverState.SCHEDULED =>
+            val scheduledExecutors = receiverTrackingInfo.scheduledExecutors.get
+            // The probability that a scheduled receiver will run in an executor is
+            // 1.0 / scheduledLocations.size
+            scheduledExecutors.map(location => location -> (1.0 / scheduledExecutors.size))
+          case ReceiverState.ACTIVE => Seq(receiverTrackingInfo.runningExecutor.get -> 1.0)
+        }
     }.groupBy(_._1).mapValues(_.map(_._2).sum) // Sum weights for each executor
 
-    val idleExecutors = (executors.toSet -- executorWeights.keys).toSeq
-    if (idleExecutors.size >= preferredNumExecutors) {
-      // If there are more than `preferredNumExecutors` idle executors, return all of them
+    val idleExecutors = executors.toSet -- executorWeights.keys
+    if (idleExecutors.nonEmpty) {
       scheduledExecutors ++= idleExecutors
     } else {
-      // If there are less than `preferredNumExecutors` idle executors, return 3 best options
-      scheduledExecutors ++= idleExecutors
-      val sortedExecutors = executorWeights.toSeq.sortBy(_._2).map(_._1)
-      scheduledExecutors ++= (idleExecutors ++ sortedExecutors).take(preferredNumExecutors)
+      // There is no idle executor. So select all executors that have the minimum weight.
+      val sortedExecutors = executorWeights.toSeq.sortBy(_._2)
+      if (sortedExecutors.nonEmpty) {
+        val minWeight = sortedExecutors(0)._2
+        scheduledExecutors ++= sortedExecutors.takeWhile(_._2 == minWeight).map(_._1)
+      } else {
+        // This should not happen since "executors" is not empty
```

Comment:
Are you talking about `executorWeights no being empty? 
And why it cannot be empty? Is it because that by the time `rescheduleReceiver` is called, all the receivers have already been scheduling by `scheduleReceivers` which means that there are some weights for at some executors?